backends/candle/src/models/nomic.rs

```diff
@@ -92,48 +95,336 @@ impl NomicBertEmbeddings {
+        })
+    }
+
+    pub fn forward(
+        &self,
+        hidden_states: &Tensor,
+        top_weights: &Tensor,
+        top_experts: &Tensor,
+    ) -> Result<Tensor> {
+        let _enter = self.span.enter();
+
+        let (bs, seq_len, hidden_size) = hidden_states.dims3()?;
+
+        let hidden_states = hidden_states.reshape(((), hidden_size))?;
+
+        let mut out = Tensor::zeros_like(&hidden_states)?;
+
+        let counts = top_experts
+            .flatten_all()?
+            .bincount(self.moe_num_experts as u32)?;
+
+        for (expert_idx, &count) in counts.iter().enumerate().take(self.moe_num_experts) {
+            if count == 0u32 {
+                continue;
+            }
+
+            let idx_top = top_experts.eq(expert_idx as f64)?.nonzero()?.t()?;
+            let idx = &idx_top.i(0)?.contiguous()?;
+            let top = &idx_top.i(1)?.contiguous()?;
+
+            let expert_out = self
+                .mlp
+                .forward(&hidden_states.index_select(idx, 0)?, expert_idx)?
+                .broadcast_mul(
+                    &top_weights
+                        .index_select(idx, 0)?
+                        .gather(&top.unsqueeze(1)?, 1)?
+                        .squeeze(1)?
+                        .unsqueeze(D::Minus1)?
+                        .to_dtype(hidden_states.dtype())?,
+                )?;
+
+            out = out.index_add(idx, &expert_out, 0)?;
+        }
+
+        let out = out.reshape((bs, seq_len, hidden_size))?;
+
+        out.broadcast_add(&self.bias)
+    }
+}
+
+pub struct NomicMoELayer {
+    router: NomicRouter,
+    experts: NomicExperts,
+
+    span: tracing::Span,
+}
+
+impl NomicMoELayer {
+    pub fn load(vb: VarBuilder, config: &NomicConfig) -> Result<Self> {
+        let router = NomicRouter::load(vb.pp("router"), config)?;
+        let experts = NomicExperts::load(vb.pp("experts"), config)?;
+
+        Ok(Self {
+            router,
+            experts,
+            span: tracing::span!(tracing::Level::TRACE, "moe"),
+        })
+    }
+
+    pub fn forward(&self, hidden_states: &Tensor) -> Result<Tensor> {
+        let _enter = self.span.enter();
+
+        let (top_weights, top_experts) = self.router.forward(hidden_states)?;
+
+        self.experts
+            .forward(hidden_states, &top_weights, &top_experts)
+    }
+}
+
+enum NomicMLP {
+    MoE(NomicMoELayer),
+    GatedMLP(NomicBertGatedMLP),
+    Mlp(NomicBertMLP),
+}
+
+impl NomicMLP {
+    pub fn load(vb: VarBuilder, index: usize, config: &NomicConfig) -> Result<Self> {
+        let use_moe = matches!(config.moe_every_n_layers, Some(n) if n > 0 && index % n == 1);
+
+        match () {
+            _ if use_moe => Ok(Self::MoE(NomicMoELayer::load(vb, config)?)),
+            _ if config.activation_function == HiddenAct::Gelu => {
+                Ok(Self::Mlp(NomicBertMLP::load(vb, config)?))
+            }
+            _ => Ok(Self::GatedMLP(NomicBertGatedMLP::load(vb, config)?)),
+        }
+    }
```

Comment:
This seems wrong.

if else sequence seems more idiomatic than empty match here.

activation_function deciding the type of MLP layers seems wrong, can't we find some better config option for it ?

Reply:
Surprisingly, NomicBert decides the type of MLP layer by activation function. Here is [the code](https://huggingface.co/nomic-ai/nomic-bert-2048/blob/main/modeling_hf_nomic_bert.py#L1635).

Well, for now, the only way I can think of is to identify the type of MLP layer by the key name in the weights, instead of relying on the activation function name (e.g. if there's `fc11.weight`, then use `NomicBertGatedMLP`). However, I think it's also not a good approach.

I'll think about it further.

Reply:
Maybe, I think it'd be better to make a PR to the nomic-ai models to add a new configuration for the type of MLP layer.

Reply:
forgot to leave a comment here. I had fixed the empty match part https://github.com/huggingface/text-embeddings-inference/pull/596/commits/95812261b26b0873277d7755dd97418ac9fa187b